[범블비] API 테스트/문서자동화 미션 제출합니다.

ddaaac opened the pull request:
잘 부탁드립니다 :)

icyMojito committed:
feat: 문서 자동화 실습

Spring Rest Docs를 이용하여 api를 문서화함

icyMojito committed:
feat: 로그인 프로세스 실습

3가지 인증 방식(basic, session, bearer)을 활용하여 로그인 기능 구현

icyMojito committed:
docs: 1단계 회원관리 기능 목록 작성

icyMojito committed:
feat: 회원가입 기능 구현 및 예외 상황 처리

icyMojito committed:
test: 회원가입 인수테스트

icyMojito committed:
docs: 1단계 회원관리 기능 목록 수정

ddaaac committed:
test: 로그인 기능 인수 테스트

ddaaac committed:
refactor: 비밀번호 확인 검증 로직 이동

ddaaac committed:
feat: 로그인 기능 구현 및 예외 상황 처리

icyMojito committed:
feat: 회원정보 조회/수정/삭제 기능 구현

icyMojito committed:
test: 회원정보 조회/수정/삭제 기능 인수테스트

icyMojito committed:
refactor: 사용하지 않는 Interceptor 제거

ddaaac committed:
feat: 회원 정보 관리 문서화

ddaaac committed:
test: JWT Token Provider 테스트

icyMojito committed:
feat: 회원가입, 로그인 문서화

icyMojito committed:
feat: 회원가입 페이지 연동

icyMojito committed:
refactor: 사용하지 않는 api 삭제

icyMojito committed:
feat: 로그인 페이지 연동

icyMojito committed:
feat: 로그인한 회원의 나의 정보 페이지 연동

icyMojito committed:
feat: ExceptionController에 처리 대상 예외 추가

icyMojito committed:
feat: 로그인한 회원의 나의 정보 수정 페이지 연동

icyMojito committed:
feat: 로그인한 회원의 탈퇴 페이지 연동

ddaaac committed:
docs: 2단계 즐겨찾기 기능 목록 작성

ddaaac committed:
test: 즐겨찾기 기능 인수 테스트

icyMojito committed:
feat: 즐겨찾기 등록/조회/삭제 기능 구현

icyMojito committed:
feat: 즐겨찾기 페이지 연동

icyMojito committed:
refactor: 즐겨찾기 경로라는 개념을 더 명확하게 표현하도록 클래스 이름 변경

icyMojito committed:
feat: 즐겨찾기 등록/조회/삭제 문서화

icyMojito committed:
refactor: FavoriteService 관련 dto 클래스 이름을 직관적으로 수정

사용하지 않는 메서드 제거

icyMojito committed:
refactor: 메서드가 처리하는 예외가 늘어남에 따라 메서드 이름을 포괄적으로 수정

icyMojito committed:
refactor: 의도가 잘 드러나도록 메서드 이름 수정

icyMojito committed:
test: 회원가입과 회원정보 관련 test 추가

icyMojito committed:
docs: 2단계 즐겨찾기 기능 목록 완료

phs1116 commented on src/main/java/wooteco/subway/domain/member/Member.java:
```diff
@@ -2,13 +2,21 @@
     private String email;
     private String name;
     private String password;
+    private List<FavoritePath> favoritePaths = new ArrayList<>();
```
즐겨찾기를 멤버에서 책임질 필요가 있을까? 라는 고민이 조금 드네요.
도메인 상으로도 둘이 memberId를 공유한다는 것 외에 컨텍스트를 엮을수 있는 부분이 없어보이고,
Member 자체의 책임이 늘어나는 경향이 있어서요. :)
Favorite에 대해서든 별도 바운디드 컨텍스트로 쪼개서 Favorite이 memberId를 알고있고, 이로 인하여 비즈니스를 풀어나갈수도 있지 않을까 싶어요. :) 한번 고민해보세요.

ddaaac replied:
Member가 FavoritePath를 가질지 말지에 대한 고민이 이번 미션을 진행하면서 가장 많이 고민했던 부분인 것 같습니다. 저는 이번 미션 요구사항에서 Favorite이 Member에 종속적이라 생각해서 처음과 같이 코드를 짰습니다(Member라는 네이밍보다 FavoriteMember가 더 적합했겠네요). Favorite이 Member 없이 별도로 호출되는 일이 요구사항 내에서는 없을 것 같아서요. 그런데 리뷰를 받고 생각해보니 이런 종속적인 관계는 요구사항 변경에 매우 취약할 것 같다는 생각이 들었습니다. 그래서 Member와 Favorite이 분리되게 코드를 변경했습니다. 

분리하고 나서 궁금한 점이 생겼습니다. Member가 삭제될 때 해당되는 Favorite들이 모두 삭제돼야 하는데 이 연관관계를 서비스에서 잡아줄 지, DB에서 외래키로 잡아줄지 고민됩니다. 아니면 안전하게 둘 다 설정해주는 게 좋을까요?

phs1116 replied:
DB 제약조건으로 처리를 하기에는 비즈니스 어플리케이션 내에서 파악하기 어려워서
서비스 내에서 처리해주는게 좋아보여요. :)

phs1116 commented on src/main/java/wooteco/subway/domain/member/Member.java:
```diff
@@ -20,10 +28,55 @@ public Member(String email, String name, String password) {
+        }
+    }
+
+    public void addFavoritePath(FavoritePath favoritePath) {
+        validateDuplication(favoritePath);
+
+        if (Objects.nonNull(favoritePath)) {
+            this.favoritePaths.add(favoritePath);
+        }
+    }
+
+    private void validateDuplication(FavoritePath favoritePath) {
+        favoritePaths.stream()
+                .filter(path -> Objects.equals(path.getSourceId(), favoritePath.getSourceId()) &&
+                        Objects.equals(path.getTargetId(), favoritePath.getTargetId()))
+                .findFirst()
+                .ifPresent(path -> {
+                    throw new DuplicatedFavoritePathException();
+                });
+    }
```
Member가 Favorite을 관리하도록 한다면, 해당 부분은 일급 컬렉션으로 분리해도 좋을 것 같네요. :)

phs1116 commented on src/main/java/wooteco/subway/domain/member/Member.java:
```diff
@@ -42,16 +95,27 @@ public String getPassword() {
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Member member = (Member) o;
+        return Objects.equals(id, member.id) &&
+                Objects.equals(email, member.email) &&
+                Objects.equals(name, member.name) &&
+                Objects.equals(password, member.password);
```
엔티티의 경우 동등성은 id만으로 식별해요. :)

http://wonwoo.ml/index.php/post/987

phs1116 commented on src/main/java/wooteco/subway/domain/path/FavoritePath.java:
```diff
@@ -0,0 +1,32 @@
+import org.springframework.data.annotation.Id;
+
+public class FavoritePath {
+	@Id
+	private Long id;
```
식별자에 대해서 Equals, hashcode 구현해주는게 좋아요. :)

phs1116 commented on src/main/java/wooteco/subway/service/favorite/FavoriteService.java:
```diff
@@ -0,0 +1,60 @@
+		List<FavoritePath> favoritePaths = member.getFavoritePaths();
+		return favoritePaths.stream()
+				.map(path -> {
+					Station source =
+							stationRepository.findById(path.getSourceId()).orElseThrow(() -> new NotExistStationException(path.getSourceId()));
+					Station target =
+							stationRepository.findById(path.getTargetId()).orElseThrow(() -> new NotExistStationException(path.getTargetId()));
+					return new FavoritePathResponse(path.getId(), StationResponse.of(source),
+					                                StationResponse.of(target));
```
stationId에 대해서 공통적인 Id들이 있을수 도 있는데요,
쿼리를 여러번 날려 네트워크 리소스를 소모하지 않고 set으로 묶어 in절로 요청해 한번에 가져와서 매핑해도 되지 않을까요? :)

phs1116 commented on src/main/java/wooteco/subway/web/ExceptionController.java:
```diff
@@ -0,0 +1,39 @@
+import wooteco.subway.web.member.InvalidAuthenticationException;
+
+@RestControllerAdvice
+public class ExceptionController {
```
Controller는 아닌지라 ControllerAdvice로 명확하게 네이밍 해주는게 좋을 것 같아요. :)

phs1116 commented on src/main/java/wooteco/subway/web/FavoriteController.java:
```diff
@@ -0,0 +1,42 @@
+	public ResponseEntity<Void> registerFavoritePath(@LoginMember Member member,
+	                                                 @RequestBody FavoritePathRequest request) {
+		favoriteService.registerPath(member, request.getSource(), request.getTarget());
+		return ResponseEntity.created(URI.create("/favorite/1")).build();
```
새로운 favorite가 생기더라도 아이디는 항상 1로 고정이네요? :)

ddaaac replied:
코드에 좀 더 신경쓰도록 하겠습니다... 😥

phs1116 reviewed (changes requested):
안녕하세요 범블비!
구현 잘해주셨어요. 👍 
피드백 몇가지 추가했는데 참고 부탁드릴게요. :)

ddaaac committed:
refactor: 리뷰 반영

- 엔티티 equals & hashcode 구현 및 id로 비교하도록 변경
- retrievePath 메서드에서 db 연결을 한번만 하도록 변경
- favoritePath 생성 시 생성된 id가 반영이 안되던 버그 수정
- Controller Advice 클래스 이름 변경

ddaaac committed:
refactor: FavoritePath를 Member와 분리

ddaaac committed:
refactor: FavoritePaths 일급 컬렉션 구현

ddaaac committed:
feat: Member 또는 Station을 삭제할 때, 관련된 FavoritePath를 삭제하는 기능 구현

phs1116 commented on src/main/java/wooteco/subway/domain/station/Station.java:
```diff
@@ -9,8 +8,7 @@
     private String name;
-    @CreatedDate
-    private LocalDateTime createdAt;
+    private LocalDateTime createdAt = LocalDateTime.now();
```
spring data jdbc auditing기능을 사용하지 않는 이유가 있을까요? :)

phs1116 commented on src/main/java/wooteco/subway/domain/station/Stations.java:
```diff
@@ -15,8 +21,13 @@ public Stations(List<Station> stations) {
     public Station extractStationById(Long stationId) {
         return stations.stream()
-                .filter(it -> it.getId() == stationId)
+                .filter(it -> Objects.equals(it.getId(), stationId))
```
get, equals외에 메시지를 통한 비교도 좋을 것 같아요. :)

phs1116 commented on src/main/java/wooteco/subway/service/member/MemberService.java:
```diff
@@ -3,49 +3,66 @@
-        memberRepository.save(member);
-    }
+	public void updateMember(Long id, UpdateMemberRequest param) {
+		Member member = memberRepository.findById(id).orElseThrow(RuntimeException::new);
```
별도 예외에 대한 메시지 없이 RuntimeException으로만 던지네요.
깜빡하신 것 같아요. :)

ddaaac replied:
미션 요구사항에 필요한 기능들을 FavoritePathController에서 `/me`를 활용하다보니 사용하지 않는 기능들이 생겼습니다. 지울까 말까 고민했는데, 사용하지 않는 기능들은 삭제하였습니다!

phs1116 commented on src/main/java/wooteco/subway/service/member/MemberService.java:
```diff
@@ -3,49 +3,66 @@
-            throw new RuntimeException("잘못된 패스워드");
-        }
+	public void deleteMember(Long id) {
+    	favoritePathService.deletePathByMember(id);
+		memberRepository.deleteById(id);
```
인덴트 맞춰주세요. :) (20~ 24번 라인도 같이요!)

phs1116 commented on src/main/java/wooteco/subway/web/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,39 @@
+			InvalidAuthenticationException.class})
+	public ExceptionResponse invalidLogin(RuntimeException e) {
+		return new ExceptionResponse(e.getMessage());
+	}
```
여기서 처리되지 않은 Exception에 대해서는 약속된 포멧(ExceptionResponse)으로 내려가지 않을 것 같은데요.
이에 대해서도 예외를 처리해주면 어떨까요?

시스템 예외의 경우에는 예외 메시지를 컨트롤 할 수 없다보니 임의의 메시지를 내려주는것도 좋을 것 같아요.
(에러로그도 찍으면 더욱 좋구요. :))

phs1116 commented on src/main/java/wooteco/subway/service/member/MemberService.java:
```diff
@@ -3,49 +3,66 @@
-        if (!member.checkPassword(param.getPassword())) {
-            throw new RuntimeException("잘못된 패스워드");
-        }
+	public void deleteMember(Long id) {
+    	favoritePathService.deletePathByMember(id);
+		memberRepository.deleteById(id);
+	}
```
deletePathByMember는 성공하고 deleteById는 실패할 케이스도 있을 것 같은데요.
트랜잭션을 묶어주는게 좋지 않을까요? :) (deleteMember(Member member) 도 같이. :))

phs1116 commented on src/main/java/wooteco/subway/web/FavoriteController.java:
```diff
@@ -0,0 +1,44 @@
+        this.favoritePathService = favoritePathService;
+    }
+
+    @PostMapping("/favorite/me")
```
컨트롤러 명 및 url을 자원의 이름과 맞춰주는게 좋지 않을까요? :)

phs1116 reviewed (changes requested):
안녕하세요, 범블비!
피드백 반영 잘 해주셨어요. 👍 
질문주신 부분에 제 생각과 몇 가지 피드백 추가했는데
참고 부탁드릴게요. :)

ddaaac committed:
refactor: 코드 정렬 및 import 최적화

ddaaac committed:
refactor: 엔티티 equals&hashcode 구현 및 createDate, lastModifiedDate를 베이스 엔티티로 분리

ddaaac committed:
refactor: getter를 활용한 메서드들을 메세지를 사용하도록 변경

ddaaac committed:
refactor: 불필요한 기능 삭제 및 예외처리, 일급 컬렉션 리팩토링

ddaaac committed:
feat: 내부 서버 오류 로깅 및 응답 처리

ddaaac committed:
feat: 서비스들에 알맞은 트랜잭션 추가

ddaaac committed:
refactor: FavoriteController 및 해당 url을 자원에 맞게 FavoritePath로 변경

ddaaac commented:
매번 중요한 부분들을 까먹네요... 😥  부족한 부분 더 말씀해주시면 반영하겠습니다!

phs1116 commented on src/main/java/wooteco/subway/jdbc/BaseEntity.java:
```diff
@@ -0,0 +1,31 @@
+import org.springframework.data.annotation.CreatedDate;
+import org.springframework.data.annotation.LastModifiedDate;
+
+public class BaseEntity {
```
BaseEntity를 분리하셨네요. 👍

phs1116 commented on src/main/java/wooteco/subway/service/path/PathService.java:
```diff
@@ -74,9 +77,9 @@ private Station extractStation(Long stationId, List<Station> stations) {
-                    .orElseThrow(RuntimeException::new);
+                .filter(it -> it.isLineStationOf(finalPreStationId, stationId))
+                .findFirst()
+                .orElseThrow(RuntimeException::new);
```
다른부분 정말 잘 처리하셨는데 깜빡 하신 듯 하네요. :)

phs1116 commented on src/main/java/wooteco/subway/web/PageController.java:
```diff
@@ -17,9 +18,14 @@ public PageController(LineService lineService, StationService stationService) {
         this.stationService = stationService;
     }
 
+    @GetMapping(value = "/admin", produces = MediaType.TEXT_HTML_VALUE)
```
PageController가 책임지고 있는 내용이 너무 많네요.
슬슬 컨트롤러도 분리해도 되지 않을까 싶어요. :)

ddaaac replied:
질문이 있습니다!
찾아보니 `WebMvcConfigure`에서 `addViewController()` 메서드를 오버라이드 하면 현재 PageController와 동일한 기능을 수행할 수 있습니다.
그런데 말씀해주신 책임의 분리를 생각하면 `PageController`를 여러 개로 나누는 것이 좋아보이는데 어떤 방법이 많이 쓰이나 궁금합니다 :)

phs1116 replied:
컨트롤러를 여러개로 나누는 방식을 많이 사용해요. :)

phs1116 commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthInterceptor.java:
```diff
@@ -22,12 +24,13 @@ public BearerAuthInterceptor(AuthorizationExtractor authExtractor, JwtTokenProvi
     public boolean preHandle(HttpServletRequest request,
                              HttpServletResponse response, Object handler) {
-        // TODO: Authorization 헤더를 통해 Bearer 값을 추출 (authExtractor.extract() 메서드 활용)
+        String credentials = authExtractor.extract(request, "bearer");
```
"bearer" 하드코딩보다는 상수로 관리하는게 더 좋을 것 같네요. :)

phs1116 reviewed (changes requested):
안녕하세요, 범블비!
피드백 반영, 리팩토링 👍 
구현 깔끔하게 잘해주셨어요.
사소한 피드백 몇가지 추가했는데 참고 부탁드려요. :)

ddaaac committed:
refactor: LineStation이 존재하지 않을 시 예외 처리 추가
- 리소스가 존재하지 않는 예외들을 ResourceNotExist 예외로 묶음

ddaaac committed:
refactor: dto에서 불필요한 기본 생성자 제거

ddaaac committed:
refactor: 원시 값 상수화

ddaaac committed:
refactor: Page Controller를 Service와 Admin으로 분리

ddaaac commented:
예외처리에 대해서 궁금한 점이 하나 생겨서 질문 남깁니다!
지금 코드에서는 클라이언트에게 보내는 예외에 대한 response에 **메세지만** 담겨 있습니다.
조금 공부를 하다보니 http 상태 코드 외에 **커스텀 코드**를 활용하는 방법도 있다고 알게 되었습니다.
이에 대해서 여러 얘기를 해봤는데 어떤 방법이 좋은 방법일지 헷갈립니다 😕 
현업에서는, 리뷰어님은 어떤 방식을 주로 사용하시는지 궁금합니다 :)

phs1116 commented:
예외처리에 대해서 궁금한 점이 하나 생겨서 질문 남깁니다!
지금 코드에서는 클라이언트에게 보내는 예외에 대한 response에 메세지만 담겨 있습니다.
조금 공부를 하다보니 http 상태 코드 외에 커스텀 코드를 활용하는 방법도 있다고 알게 되었습니다.
이에 대해서 여러 얘기를 해봤는데 어떤 방법이 좋은 방법일지 헷갈립니다 😕
현업에서는, 리뷰어님은 어떤 방식을 주로 사용하시는지 궁금합니다 :)

----

http status code로는 클라이언트에게 모든 상황에 대해 알려주기 어려울수도 있어요.
비즈니스 상황에 대해서 클라이언트에서 별도 처리를 한다고 가정한다면요.
말씀하신대로 현재 쓰고있는 ResponseEntity 외에 별도 Response 공통 제네릭 클래스를 만들어서,
데이터와 코드값을 전달하기도 해요.
아래 형태가 될 수 있겠죠?


```
public class ResponseDTO<T> {
	private String responseCode;
	private String responseMessage;
	private T content;
}
```

phs1116 commented on src/main/java/wooteco/subway/exceptions/ResourceNotExistException.java:
```diff
@@ -0,0 +1,7 @@
+package wooteco.subway.exceptions;
+
+public class ResourceNotExistException extends RuntimeException {
```
예외를 한번 더 추상화 하셨네요. 👍 
하신 것처럼 비즈니스 케이스를 정리해서 이런식으로 추상화 하기도 해요.

phs1116 reviewed (approved):
안녕하세요, 범블비!
피드백 반영 잘해주셨어요. 👍 
이번 단계 머지할게요!
질문주신 사항에 대한 답변도 추가했는데 한번 참고 부탁드려요. :)
개발하시느라 수고 많으셨어요!
